openg2p_registration/models/registration_service.py

```diff
@@ -9,10 +10,77 @@ class RegistrationService(models.Model):
+            odk_data["program_ids"][len(odk_data["program_ids"]) - 1]
+        )
+        program_obj = self.env["openg2p.program"].search(
+            [("id", "=", current_program_id)]
+        )
         self.renaming_submission_data_fields(temp, odk_data)
-        self.create_fields_for_registration(temp)
+        autodedup_field = program_obj.autodedup_field
+        action = program_obj.action
+        stage_name = program_obj.stage_name
+        if autodedup_field == "kyc":
+            kycdup_list = (
+                self.env["openg2p.registration"]
+                .search([("kyc_id", "=", temp["bban"])])
+                .ids
+            )
 
-    def create_fields_for_registration(self, temp):
+            if len(kycdup_list) != 0:
+                if action == "merge":
+                    # merging the existing ones with new one and passing the existing_id and current_id as arguments
+                    self.merge_registrations(temp, kycdup_list[0])
+                elif action == "del_old":
+                    # deleting the old registration
+                    duplicate_kyc = self.env["openg2p.registration"].search(
+                        [("kyc_id", "=", temp["bban"])]
+                    )
+                    duplicate_kyc.active = False
+                    # deleting the old beneficiary
+                    duplicate_kyc_bene = self.env["openg2p.beneficiary"].search(
+                        [("kyc_id", "=", temp["bban"])]
+                    )
+                    duplicate_kyc_bene.active = False
+                    # creating registration for new record
+                    self.create_fields_for_registration(temp, stage_name)
+                elif action == "del_new":
+                    print("kyc and delete new")
+                    # deleting the current record
+            else:
+                self.create_fields_for_registration(temp, stage_name)
+        elif autodedup_field == "ext_id":
+            externaldup_list = (
+                self.env["openg2p.registration"]
+                .search([("external_id", "=", temp["new_emis_code"])])
+                .ids
+            )
+            if len(externaldup_list) != 0:
+                if action == "merge":
+                    # merging the existing ones with new one and passing the existing_id and current_id as arguments
+                    self.merge_registrations(temp, externaldup_list[0])
+                elif action == "del_old":
+                    # deleting the old registration
+                    duplicate_external = self.env["openg2p.registration"].search(
+                        [("external_id", "=", temp["new_emis_code"])]
+                    )
+                    duplicate_external.active = False
+                    # deleting the old beneficiary
+                    duplicate_external_bene = self.env["openg2p.beneficiary"].search(
+                        [("external_id", "=", temp["new_emis_code"])]
```

Comment:
EMIS code has nothing to do with autodedup other than the part that it's part of ODK submission and is used to populate external ID field of registration.

Reply:
I am checking duplicates before registration is created so for this reason I am using that field. Here temp is a dictionary containing the odk submissions